src/bindings/python/flux/job.py

```diff
@@ -267,6 +267,155 @@ def wait(flux_handle, jobid=lib.FLUX_JOBID_ANY):
     return future.get_status()
 
 
+class EventLogEvent:
+    """
+    wrapper class for a single KVS EventLog entry
+    """
+
+    def __init__(self, event):
+        """
+        "Initialize from a string or dict eventlog event
+        """
+        self.event = event
+        if type(event) is str:
+            self.event = json.loads(event)
+        self._name = self.event["name"]
+        self._timestamp = self.event["timestamp"]
+        self._context = {}
+        if "context" in self.event:
+            self._context = self.event["context"]
+
+    def __str__(self):
+        return "{0.timestamp:<0.5f}: {0.name} {0.context}".format(self)
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def timestamp(self):
+        return self._timestamp
+
+    @property
+    def context(self):
+        return self._context
+
+
+class JobEventWatchFuture(Future):
+    """
+    A future returned from job.event_watch(). The future is
+    initially "empty" but queues EventLogEvent results for later
+    consumption by caller via .get_event() method.
+    """
+
+    def __init__(self, handle):
+        future = RAW.flux_future_create(ffi.NULL, ffi.NULL)
+        super().__init__(future)
+        self.pimpl.set_flux(handle)
+        self.results = collections.deque()
+
+    def fulfill(self, result):
+        self.results.append(result)
+        self.pimpl.fulfill(ffi.NULL, ffi.NULL)
+
+    def fulfill_error(self, errno, errmsg):
+        self.pimpl.fulfill_error(errno, errmsg)
+
+    def get_event(self):
+        self.wait_for()
+        if self.results:
+            return self.results[0]
+        result = self.pimpl.get(ffi.NULL)
+        if result == -1:
+            raise OSError(ffi.errno, self.error_string())
+        return result
+
+    def reset(self):
+        super().reset()
+        if self.results:
+            self.results.popleft()
+
+
+def event_watch_async(flux_handle, jobid, name="*", eventlog="eventlog"):
```

Comment:
This is really nice!  For anyone that is familiar with event-driven development and Flux's futures, this is perfect!

For people more familiar with Python than Flux, it might also be nice to have an `event_watch` function that is a generator so that someone can easily write a for loop over the events:

```python
for event in job.event_watch(fh, jobid):
   print("Event: {}".format(event))
```

That loop has the downside of only progressing as fast as the job progresses through its states (that may surprise some users), but I could see how it would be useful (e.g., in `event_wait`).

Reply:
The problem is that `event_watch_async` cannot be used synchronously at the moment. See the bottom of the PR description.

This is the right feedback though. I was looking for comments on what interfaces would be most useful.